File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-7.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        auto dummy = new ListNode(0, head);
        
        auto slow = dummy;
        auto fast = dummy;
        
        for (int i = 0; i <= n; ++i)
            fast = fast->next;
        
        while (fast) {
            slow = slow->next;
            fast = fast->next;
        }

        auto node = slow->next;
        slow->next = node->next;
        delete node;

        head = dummy->next;
        delete dummy;
        return head;
    }
};

File: Data Structures & Algorithms/copy-linked-list-with-random-pointer/submission-7.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        dummy = Node(0)
        curr, currCopy = head, dummy
        copies = {}
        while curr:
            currCopy.next = Node(curr.val)
            copies[curr] = currCopy.next
            curr, currCopy = curr.next, currCopy.next
        curr, currCopy = head, dummy.next
        while currCopy:
            currCopy.random = copies[curr.random] if curr.random else None
            curr, currCopy = curr.next, currCopy.next
        return dummy.next